app/services/wakes/pageview_count_updater_service.rb

```diff
@@ -0,0 +1,69 @@
+class Wakes::PageviewCountUpdaterService
+  def initialize(location)
+    @location = location
+  end
+
+  attr_reader :location
+
+  def update_pageview_count
+    if pageviews_since_last_update > 0
+      update_location_pageview_count && update_resource_aggregate_count && update_wakeable_aggregate_count
+    else
+      false
+    end
+  ensure
+    # Yes, this will often result in two update operations
+    # Right now, code expressiveness and comprehension seem more important than performance
+    location.update(:pageview_count_checked_at => Time.zone.now)
+  end
+
+  def update_location_pageview_count
+    location.update(:pageview_count => location.pageview_count.to_i + pageviews_since_last_update,
+                    :pageview_count_updated_through => end_date)
+  end
+
+  # If a resource's locations differ only by their query string, then this method will not work because
+  # of the way the GA regex is working. The old code handled this case. It is not yet certain whether
+  # we need wakes to handle it as well, because of how wakes stores its paths.
+  #
+  # This is the core of the old logic: https://gist.github.com/benhutton/c685aa1f8942553a9745
+  def update_resource_aggregate_count
+    resource = location.resource
+    count = resource.locations.sum("COALESCE((wakes_locations.document ->> 'pageview_count')::int, 0)")
+    resource.update(:pageview_count => count)
+  end
+
+  def update_wakeable_aggregate_count
+    return true unless wakeable = location.resource.wakeable
+    return true unless wakeable.respond_to?(:pageview_count=)
```

Comment:
@benhutton This is mostly a matter of personal preference, but I would write the above block as:
```
wakeable = location.resource.wakeable
return true unless wakeable && wakeable.respond_to?(:pageview_count=)
```